### PR TITLE
Allow registration of custom buttons

### PR DESCRIPTION
**Pull request by josephfusco**
Solves #31

This also registers a new share button as an example, using [@wordpress/icons](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-icons/) package 🎉 

- [x] encode using LZString
- [x] decode using LZString
- [x] Implement clipboard functionality & ~notify user (currently console.log)~ notification system is out of scope of this PR

custom buttons

Props to @jasonbahl for the [POC](https://github.com/wp-graphql/wpgraphql-ide/pull/42) 🙌🏻

### CODE REVIEW

File: src/components/Editor.jsx [hunk 1]
```diff
@@ -20,6 +35,102 @@ const fetcher = async ( graphQLParams ) => {
...
+
+	return (
+		<ToolbarButton
+			onClick={ generateShareLink }
+			label="Share current document"
+		>
+			<Icon
+				icon={ external }
+				style={ {
+					fill: 'hsla(var(--color-neutral), var(--alpha-tertiary))',
```
Comment (josephfusco) [inline src/components/Editor.jsx:116]: These exact css variables are what the other icons are using. Something to consider abstracting into it's own component for any future theming effort

File: wpgraphql-ide.php [hunk 2]
```diff
@@ -206,11 +204,12 @@ function enqueue_react_app_with_styles(): void {
...
-			'graphqlEndpoint'    => trailingslashit( site_url() ) . 'index.php?' . \WPGraphQL\Router::$route,
-			'rootElementId'      => WPGRAPHQL_IDE_ROOT_ELEMENT_ID,
-			'context'            => $app_context,
-			'isDedicatedIdePage' => is_dedicated_ide_page(),
+			'nonce'               => wp_create_nonce( 'wp_rest' ),
+			'graphqlEndpoint'     => trailingslashit( site_url() ) . 'index.php?' . \WPGraphQL\Router::$route,
+			'rootElementId'       => WPGRAPHQL_IDE_ROOT_ELEMENT_ID,
+			'context'             => $app_context,
+			'isDedicatedIdePage'  => is_dedicated_ide_page(),
+			'dedicatedIdeBaseUrl' => get_dedicated_ide_base_url(),
```
Comment (josephfusco) [inline wpgraphql-ide.php:212]: Hard to tell with this diff, but `dedicatedIdeBaseUrl` is new.

**Review by jasonbahl — CHANGES_REQUESTED** [review]
see my comment about the `query` query_arg. 

Then, I think my P.O.C might have been a bit too naiive. 

I think we might want to abstract a few things to make this more "error proof" so to speak. 

For example, instead of requiring each Button to import and know about the `<ToolbarButton>`, `<Icon>` and `<VisuallyHidden>` components, we could include those elements within our `object.entries.map()` and render the component based on "props" in the `toolbarButtons`. 

for example, when I do `add_menu_page()` in WordPress, I pass along strings of data, like: `page_title`, `menu_title`, `menu_slug`, etc and the UI is built for me. 

I think ultimately we probably want that type of DX. 

The buttons will all be given a consistent look, a tooltip, etc. . .all you need to provide is the label, the icon, the tooltip description and an onClick callback. 

So the end user might do something like: 

```js
registerGraphiqlToolbarButton({
  name: "share",
  label: "Share current document",
  icon: () => { return <SomeIcon /> },
  description: "The tooltip description",
  onClick: () => { // this executes when the button is clicked }
});
```

I, of course, never mentioned that this is the type of API I would like to see, so definitely no reason you should have expected to provide it, but I think ultimately that's where this should go.

Similar to `register_graphql_field`, `register_post_type`, `register_taxonomy`, etc

I'm working on a similar style "registry" API for the "plugins" to add the "Help" plugin, so hopefully that can shed light on the implementation details I have in mind a bit more.

**Comment by jasonbahl** [comment]
re: `registerToolbarButton` (or whatever we call it), I think that in the redux store we should have a list of buttons. Then functions like `registerToolbarButton` or `deregisterToolbarButton` will add to / remove from that list. 

Then the list, after any addition/removal has been done, is rendered to the toolbar.

File: src/components/Editor.jsx [hunk 3]
```diff
@@ -20,6 +35,102 @@ const fetcher = async ( graphQLParams ) => {
...
+const ShareDocumentButton = () => {
+	const { dedicatedIdeBaseUrl } = WPGRAPHQL_IDE_DATA;
+	const { queryEditor } = useEditorContext();
+	const [ shareUrl, setShareUrl ] = useState( '' );
+
+	const generateShareLink = () => {
+		const document = queryEditor?.getValue() ?? '';
+		const compressedQuery =
+			LZString.compressToEncodedURIComponent( document );
+		const fullUrl = `${ dedicatedIdeBaseUrl }&query=${ compressedQuery }`;
```
[the remark this answers is not in the document]
Comment (jasonbahl) [reply src/components/Editor.jsx:100]: ah, I see in the PR comment you said `@todo: decode using LZString` my b.

**Comment by josephfusco** [comment]
💯 I am with you on abstracting things out more, but in a follow up PR.

Related: https://github.com/wp-graphql/wpgraphql-ide/pull/60#discussion_r1506932766

What you have laid out makes good sense to me.
Mind creating a separate ticket to improve the developer interface? Thank you 🙌🏻

**Comment by jasonbahl** [comment]
@josephfusco I opened a follow-up issue here: https://github.com/wp-graphql/wpgraphql-ide/issues/62

File: src/components/Editor.jsx (hunk 1 above)
Comment (jasonbahl) [reply src/components/Editor.jsx:116]: Good callout. 

I included this in my PR for the "help" page / `registerPlugin` PR (#61). . .(although not fully abstracted yet)

**Comment by jasonbahl** [comment]
@josephfusco looks like a test needs updated to reflect the change to the functionality (moving from multiple supported url query params to 1 hashed param)

File: src/components/toolbarButtons/ShareDocumentButton.jsx [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+		const queryParamShareObject = {
+			query: compressedQuery,
+		};
+		const compressedQueryParamShareObject =
+			LZString.compressToEncodedURIComponent(
+				JSON.stringify( queryParamShareObject )
+			);
+
+		const fullUrl = `${ dedicatedIdeBaseUrl }&wpgraphql_ide=${ compressedQueryParamShareObject }`;
```
Comment (jasonbahl) [inline src/components/toolbarButtons/ShareDocumentButton.jsx:40]: To make this easier to test, we should consider abstracting the couple lines above into a function that returns an encoded string for use in the url. 

I also think the double use of LZString is probably unnecessary (but might want to confirm that). 

I think passing an object to a function that stringifies it then lets LZString do it's thing once is _probably_ enough (but should confirm)?

Then, in the tests we could do something like: 

```
import { getEncodedQueryParams } from '/wherever/';

const encodedQueryParams = getEncodedQueryParams( {
  query: `{posts{nodes{id}}}`
} );

await goToPage( http://localhost:8888/wp-admin/index.php?wpgraphql_ide={ getEncodedQueryParams() }`
```

**Review by jasonbahl — APPROVED** [review]

**Review by jasonbahl — APPROVED** [review]

**Pull request merged**